javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java

```diff
@@ -334,26 +334,52 @@ public static boolean containsSameByName(
-            new Comparator<AnnotationMirror>() {
-                @Override
-                public int compare(AnnotationMirror a1, AnnotationMirror a2) {
-                    // AnnotationMirror.toString() prints the elements of an annotation in the
-                    // order in which they were written. So, use areSame to check for equality.
-                    if (AnnotationUtils.areSame(a1, a2)) {
-                        return 0;
-                    }
-
-                    String n1 = a1.toString();
-                    String n2 = a2.toString();
-
-                    // Because the AnnotationMirror.toString prints the annotation as it appears
-                    // in source code, the order in which annotations of the same class are
-                    // sorted may be confusing.  For example, it might order
-                    // @IntRange(from=1, to=MAX) before @IntRange(to=MAX,from=0).
-                    return n1.compareTo(n2);
-                }
-            };
+    /**
+     * Provide ordering for {@link AnnotationMirror}s. AnnotationMirrors are first compared by their
+     * fully qualified names, then by their element values, in ordered of the name of the element.
+     *
+     * @param a1 the first annotation
+     * @param a2 the second annotation
+     * @return an ordering over AnnotationMirrors based on their name and values
+     */
+    public static int compareAnnotationMirrors(AnnotationMirror a1, AnnotationMirror a2) {
+        if (!AnnotationUtils.areSameByName(a1, a2)) {
+            return annotationName(a1).compareTo(annotationName(a2));
+        }
+        if (AnnotationUtils.sameElementValues(a1, a2)) {
+            return 0;
+        }
+        // The annotations have the same name, but different values, so compare values.
+        Map<? extends ExecutableElement, ? extends AnnotationValue> vals1 = a1.getElementValues();
+        Map<? extends ExecutableElement, ? extends AnnotationValue> vals2 = a2.getElementValues();
+        Set<ExecutableElement> sortedElements =
+                new TreeSet<>(Comparator.comparing(ElementUtils::getSimpleName));
+        sortedElements.addAll(
+                ElementFilter.methodsIn(a1.getAnnotationType().asElement().getEnclosedElements()));
+
+        for (ExecutableElement method : sortedElements) {
+            AnnotationValue value1 = vals1.get(method);
+            AnnotationValue value2 = vals2.get(method);
+            if (value1 == null) {
+                value1 = method.getDefaultValue();
+            }
+            if (value2 == null) {
+                value2 = method.getDefaultValue();
+            }
+            if (value1 == value2) {
+                continue;
+            } else if (value1 == null) {
+                return -1;
+            } else if (value2 == null) {
+                return 1;
+            }
+            int result = value1.getValue().toString().compareTo(value2.getValue().toString());
```

Comment:
Instead of doing the String comparison here, wouldn't it be nicer to change
`boolean sameAnnotationValueValue(Object val1, Object val2)` https://github.com/typetools/checker-framework/blob/master/javacutil/src/main/java/org/checkerframework/javacutil/AnnotationUtils.java#L558
to be a comparator? That is, make a version that returns `int` and the `boolean` version compares against 0.